tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/lhlo_ops.td

```diff
@@ -1386,4 +1386,191 @@ def TerminatorOp :
     [{ build($_builder, $_state, llvm::None, operands, llvm::None); }]>];
 }
 
+
+def LHLO_D2HOp: LHLO_Op<"d2h", []> {
+  let summary = "LHLO D2H operator";
+  let description = [{
+    Copy operand to output. operand is in device memory. output is in 
+    host memory
+  }];
+  let arguments = (ins
+    Arg<LHLO_Buffer, "", [MemRead]>:$operand,
+    Arg<LHLO_Buffer, "", [MemWrite]>:$output
+  );
```

Comment:
It seems that you introduce the concept of host and device memory at the memref level, this does not exist yet in LMHLO right?

This is worth a thread on the mailing list in itself, I could see us not use `LHLO_Buffer` but `LHLO_HostBuffer` for example and require it to be a memref with a specific memory space attached to it.

Reply:
Will add H2D/D2H to a new dialect. Remove them here.